crates/proof-of-sql/src/base/database/table_ref.rs

```diff
@@ -1,55 +1,125 @@
-    resource_id: ResourceId,
+    schema_name: Option<Ident>,
+    table_name: Ident,
 }
 
 impl TableRef {
-    /// Creates a new table reference from a resource id
+    /// Creates a new table reference from schema and table names.
+    /// If the schema name is empty or None, only the table name is used.
     #[must_use]
-    pub fn new(resource_id: ResourceId) -> Self {
-        Self { resource_id }
+    pub fn new(schema_name: impl AsRef<str>, table_name: impl AsRef<str>) -> Self {
+        let schema = schema_name.as_ref();
+        let table = table_name.as_ref();
+
+        Self {
+            schema_name: if schema.is_empty() {
+                None
+            } else {
+                Some(Ident::new(schema.to_string()))
+            },
+            table_name: Ident::new(table.to_string()),
+        }
     }
 
     /// Returns the identifier of the schema
+    /// # Panics
+    #[allow(clippy::get_first)]
     #[must_use]
-    pub fn schema_id(&self) -> Ident {
-        self.resource_id.schema().into()
+    pub fn schema_id(&self) -> Option<&Ident> {
+        self.schema_name.as_ref()
     }
 
+    /// # Panics
     /// Returns the identifier of the table
     #[must_use]
-    pub fn table_id(&self) -> Ident {
-        self.resource_id.object_name().into()
+    pub fn table_id(&self) -> &Ident {
+        &self.table_name
+    }
+
+    /// Creates a new table reference from an optional schema and table name.
+    #[must_use]
+    pub fn with_optional_schema(schema_name: Option<&str>, table_name: &str) -> Self {
+        Self {
+            schema_name: schema_name.map(|s| Ident::new(s.to_string())),
+            table_name: Ident::new(table_name.to_string()),
+        }
+    }
+
+    /// Creates a `TableRef` directly from `Option<Ident>` for schema and `Ident` for table.
+    #[must_use]
+    pub fn with_idents(schema_name: Option<Ident>, table_name: Ident) -> Self {
+        Self {
+            schema_name,
+            table_name,
+        }
+    }
+
+    /// Creates a `TableRef` from a slice of string components.
+    /// Expects one component for table or two components for schema and table.
+    /// Panics if the number of components is invalid.
+    /// # Panics
+    #[must_use]
+    pub fn from<S: AsRef<str>>(components: &[S]) -> Self {
+        match components.len() {
+            1 => Self::with_optional_schema(None, components[0].as_ref()),
+            2 => Self::with_optional_schema(Some(components[0].as_ref()), components[1].as_ref()),
+            _ => {
+                panic!("TableRef must have either one (table) or two (schema and table) components")
+            }
+        }
     }
 
-    /// Returns the underlying resource id of the table
+    /// Creates a `TableRef` from a dot-separated string.
     #[must_use]
-    pub fn resource_id(&self) -> ResourceId {
-        self.resource_id
+    pub fn from_string(s: &str) -> Self {
+        let components: Vec<_> = s.split('.').map(ToString::to_string).collect();
```

Comment:
This needs to return a Result since a general string e.g `i.am.not.a.legitimate.table_name` may not be a legit table name.

Reply:
Hmm why not implement `TryFrom` instead?